File: NextcloudTalk/Chat/Chat cells/SystemMessageTableViewCell.m
/**
 * SPDX-FileCopyrightText: 2020 Nextcloud GmbH and Nextcloud contributors
 * SPDX-License-Identifier: GPL-3.0-or-later
 */

#import "SystemMessageTableViewCell.h"

#import "NCAppBranding.h"

#import "NextcloudTalk-Swift.h"

static CGFloat kChatCellDateLabelWidth      = 40.0;

@interface SystemMessageTableViewCell () <UITextFieldDelegate>
@property (nonatomic, assign) BOOL didCreateSubviews;
@end

@implementation SystemMessageTableViewCell

- (instancetype)initWithStyle:(UITableViewCellStyle)style reuseIdentifier:(NSString *)reuseIdentifier
{
    self = [super initWithStyle:style reuseIdentifier:reuseIdentifier];
    if (self) {
        self.selectionStyle = UITableViewCellSelectionStyleNone;
        self.backgroundColor = [UIColor systemGroupedBackgroundColor];
    }
    return self;
}

- (void)configureSubviews
{
    [self.contentView addSubview:self.dateLabel];
    [self.contentView addSubview:self.bodyTextView];
    [self.contentView addSubview:self.collapseButton];

    NSDictionary *views = @{@"dateLabel": self.dateLabel,
                            @"bodyTextView": self.bodyTextView,
                            @"collapseButton" : self.collapseButton
                            };
    
    NSDictionary *metrics = @{@"dateLabelWidth": @(kChatCellDateLabelWidth),
                              @"avatarGap": @50,
                              @"right": @10,
                              @"left": @5
                              };
    
    [self.contentView addConstraints:[NSLayoutConstraint constraintsWithVisualFormat:@"H:|-avatarGap-[bodyTextView]-[dateLabel(>=dateLabelWidth)]-right-|" options:0 metrics:metrics views:views]];
    [self.contentView addConstraints:[NSLayoutConstraint constraintsWithVisualFormat:@"H:|-left-[collapseButton(40)]-left-[bodyTextView]-[dateLabel(>=dateLabelWidth)]-right-|" options:NSLayoutFormatAlignAllCenterY metrics:metrics views:views]];
    [self.bodyTextView.centerYAnchor constraintEqualToAnchor:self.contentView.centerYAnchor].active = YES;

    self.didCreateSubviews = YES;
}

- (void)prepareForReuse
{
    [super prepareForReuse];

    if (!self.didCreateSubviews) {
        return;
    }
    
    self.selectionStyle = UITableViewCellSelectionStyleNone;
    self.backgroundColor = [UIColor systemGroupedBackgroundColor];

    self.bodyTextView.text = @"";
    self.dateLabel.text = @"";
}

#pragma mark - Getters

- (MessageBodyTextView *)bodyTextView
{
    if (!_bodyTextView) {
        _bodyTextView = [MessageBodyTextView new];
        _bodyTextView.dataDetectorTypes = UIDataDetectorTypeNone;
    }
    return _bodyTextView;
}

- (UILabel *)dateLabel
{
    if (!_dateLabel) {
        _dateLabel = [UILabel new];
        _dateLabel.textAlignment = NSTextAlignmentRight;
        _dateLabel.translatesAutoresizingMaskIntoConstraints = NO;
        _dateLabel.backgroundColor = [UIColor clearColor];
        _dateLabel.userInteractionEnabled = NO;
        _dateLabel.numberOfLines = 1;
        _dateLabel.font = [UIFont preferredFontForTextStyle:UIFontTextStyleFootnote];
        _dateLabel.textColor = [UIColor secondaryLabelColor];
    }
    return _dateLabel;
}

- (UIButton *)collapseButton
{
    if (!_collapseButton) {
        _collapseButton = [[UIButton alloc] initWithFrame:CGRectMake(0, 0, 40, 20)];
        _collapseButton.translatesAutoresizingMaskIntoConstraints = NO;
        [_collapseButton addTarget:self action:@selector(collapseButtonPressed) forControlEvents:UIControlEventTouchUpInside];
        [_collapseButton setImage:[UIImage systemImageNamed:@"rectangle.arrowtriangle.2.inward"] forState:UIControlStateNormal];
        _collapseButton.tintColor = [UIColor tertiaryLabelColor];
    }
    return _collapseButton;
}

- (void)collapseButtonPressed
{
    [self.delegate cellWantsToCollapseMessagesWithMessage:self.message];
}

- (void)setupForMessage:(NCChatMessage *)message
{
    self.collapseButton.hidden = (message.isCollapsed || message.collapsedMessages.count == 0);

    // If the message is not visible, we don't need to setup this cell
    if (message.isCollapsed && message.collapsedBy) {
        return;
    }

    if (!self.didCreateSubviews) {
        [self configureSubviews];
    }
    
    self.bodyTextView.attributedText = message.systemMessageFormat;
    self.messageId = message.messageId;
    self.message = message;
    
    if (!message.isGroupMessage && !(message.isCollapsed && message.collapsedBy > 0)) {
        NSDate *date = [[NSDate alloc] initWithTimeIntervalSince1970:message.timestamp];
        self.dateLabel.text = [NCUtils getTimeFromDate:date];
    }

    if (!message.isCollapsed && (message.collapsedBy > 0 || message.collapsedMessages.count > 0)) {
        self.backgroundColor = [UIColor tertiarySystemFillColor];
    } else {
        self.backgroundColor = [UIColor systemGroupedBackgroundColor];
    }

    if (message.collapsedMessages.count > 0) {
        self.selectionStyle = UITableViewCellSelectionStyleDefault;
    } else {
        self.selectionStyle = UITableViewCellSelectionStyleNone;
    }
}

@end

File: NextcloudTalk/Chat/Chat cells/SystemMessageTableViewCell.swift
//
// SPDX-FileCopyrightText: 2025 Nextcloud GmbH and Nextcloud contributors
// SPDX-License-Identifier: GPL-3.0-or-later
//

protocol SystemMessageTableViewCellDelegate: AnyObject {
    func cellWantsToCollapseMessages(with message: NCChatMessage)
}

class SystemMessageTableViewCell: UITableViewCell {

    public weak var delegate: SystemMessageTableViewCellDelegate?

    public var messageId: Int = 0
    public var message: NCChatMessage?

    private var didCreateSubviews = false

    public static let identifier = "SystemMessageCellIdentifier"

    public lazy var bodyTextView = {
        let textView = MessageBodyTextView()
        textView.dataDetectorTypes = []

        return textView
    }()

    public lazy var dateLabel = {
        let label = UILabel()
        label.textAlignment = .right
        label.translatesAutoresizingMaskIntoConstraints = false
        label.backgroundColor = .clear
        label.isUserInteractionEnabled = false
        label.numberOfLines = 1
        label.font = .preferredFont(forTextStyle: .footnote)
        label.textColor = .secondaryLabel

        return label
    }()

    private lazy var collapseButton: UIButton = {
        let button = UIButton(frame: .init(x: 0, y: 0, width: 40, height: 20))

        button.translatesAutoresizingMaskIntoConstraints = false
        button.tintColor = .tertiaryLabel
        button.setImage(UIImage(systemName: "rectangle.arrowtriangle.2.inward"), for: .normal)

        button.addAction { [weak self] in
            guard let self, let message else { return }

            self.delegate?.cellWantsToCollapseMessages(with: message)
        }

        return button
    }()

    override init(style: UITableViewCell.CellStyle, reuseIdentifier: String?) {
        super.init(style: style, reuseIdentifier: reuseIdentifier)

        self.selectionStyle = .none
        self.backgroundColor = .systemBackground
    }

    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    override func prepareForReuse() {
        super.prepareForReuse()

        guard self.didCreateSubviews else { return }

        self.selectionStyle = .none
        self.backgroundColor = .systemBackground

        self.bodyTextView.text = ""
        self.dateLabel.text = ""
    }

    public func setup(for message: NCChatMessage) {
        // swiftlint:disable:next empty_count
        self.collapseButton.isHidden = (message.isCollapsed || message.collapsedMessages.count == 0)

        // If the message is not visible, we don't need to setup this cell
        if message.isCollapsed && message.collapsedBy != nil {
            return
        }

        if !self.didCreateSubviews {
            self.configureSubviews()
        }

        self.bodyTextView.attributedText = message.systemMessageFormat
        self.messageId = message.messageId
        self.message = message

        if !message.isGroupMessage && !(message.isCollapsed && message.collapsedBy != nil) {
            let date = Date(timeIntervalSince1970: TimeInterval(message.timestamp))

            self.dateLabel.text = NCUtils.getTime(fromDate: date)
        }

        // swiftlint:disable:next empty_count
        if !message.isCollapsed && (message.collapsedBy != nil || message.collapsedMessages.count > 0) {
            self.backgroundColor = .tertiarySystemFill
        } else {
            self.backgroundColor = .systemBackground
        }

        // swiftlint:disable:next empty_count
        if message.collapsedMessages.count > 0 {
            self.selectionStyle = .default
        } else {
            self.selectionStyle = .none
        }
    }

    func configureSubviews() {
        self.contentView.addSubview(self.dateLabel)
        self.contentView.addSubview(self.bodyTextView)
        self.contentView.addSubview(self.collapseButton)

        let views = [
            "dateLabel": self.dateLabel,
            "bodyTextView": self.bodyTextView,
            "collapseButton": self.collapseButton
        ]

        let metrics = [
            "dateLabelWidth": 40,
            "avatarGap": 50,
            "right": 10,
            "left": 5
        ]

        self.contentView.addConstraints(NSLayoutConstraint.constraints(withVisualFormat: "H:|-avatarGap-[bodyTextView]-[dateLabel(>=dateLabelWidth)]-right-|", metrics: metrics, views: views))
        self.contentView.addConstraints(NSLayoutConstraint.constraints(withVisualFormat: "H:|-left-[collapseButton(40)]-left-[bodyTextView]-[dateLabel(>=dateLabelWidth)]-right-|", metrics: metrics, views: views))

        self.bodyTextView.centerYAnchor.constraint(equalTo: self.contentView.centerYAnchor).isActive = true
        self.dateLabel.centerYAnchor.constraint(equalTo: self.contentView.centerYAnchor).isActive = true
        self.collapseButton.centerYAnchor.constraint(equalTo: self.contentView.centerYAnchor).isActive = true

        self.didCreateSubviews = true
    }

}
